cmake/module/AddTargetMacros.cmake

```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2024-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+include_guard(GLOBAL)
+
+# Functions in this module are drop-in replacements for CMake's add_library
+# and add_executable functions. They are mandatory for use in the Bitcoin
+# Core project, except for imported and interface libraries.
```

Comment:
Yea, it's not ectually clear why we need all this extra machinery to do this, and it's more non-standard stuff we'll have to maintain. I would also assume we can have a better interface/encapsulation, than a function that is always called with the same arguments, otherwise it can just be missued.

Reply:
My understanding is that this is needed if we want to append to the CXXFLAGS when invoking CMake, in order to override only some options, but not all. E.g. if it would use internally "-Wfoo -Wbar" we don't want to drop/override this completely but we want to append "-Wno-foo" so that it becomes "-Wfoo -Wbar -Wno-foo". Am I right?

This implementation looks quite complicated to me but I couldn't come up with a simpler one. First I tried to use the global [add_compile_options()](https://cmake.org/cmake/help/latest/command/add_compile_options.html), but that is already out-lawed via `cmake/module/WarnAboutGlobalProperties.cmake`. Then I tried to append the extra flags to `core_interface` by something like:

```cmake
cmake_language(EVAL CODE "
  cmake_language(DEFER
    DIRECTORY ${PROJECT_SOURCE_DIR}
    CALL target_compile_options core_interface INTERFACE ${EXTRA_CPPFLAGS} $<$<COMPILE_LANGUAGE:CXX>:${EXTRA_CXXFLAGS}>
  )
")
```
but to my annoyance the extra flags appear before all the `-Werror -W...` flags :angry:

Reply:
> E.g. if it would use internally "-Wfoo -Wbar" we don't want to drop/override this completely but we want to append "-Wno-foo" so that it becomes "-Wfoo -Wbar -Wno-foo". Am I right?

Correct.

> This implementation looks quite complicated to me but I couldn't come up with a simpler one. First I tried to use the global [add_compile_options()](https://cmake.org/cmake/help/latest/command/add_compile_options.html), but that is already out-lawed via `cmake/module/WarnAboutGlobalProperties.cmake`.

It is out-lawed for a reason :)
Directory-level properties are quite error-prone and hard to reason about in large project like Bitcoin Core.

> Then I tried to append the extra flags to `core_interface` by something like...

It is still possible combine extra flags into the `core_interface` library, and use it in these invocations. But, again, this approach is a bit error-prone as it it requires an extra care about the order of appending all other flags to the `core_interface` library.